### ISSUE TITLE
Sort songs by style

### ISSUE THREAD

**Issue by andrasfuchs** [ytmdesktop/ytmdesktop#745]
- [x] I understand that **YTMDesktop have NO affiliation with Google or YouTube**.
- [x] I verified that there is no open issue for the same subject.

**Is your feature request related to a problem? Please describe.**
This might not be possible in the client only, but it would be excellent to have a special sorting method in the playlist, where the similar songs are close to each other.

Similarly to the colors of a rainbow, the playlist would start with one style and gradually change throughout the whole playlist, playing similar song after each other, but gradually changing the style. If you had enough similar songs your playlist could have a "mood-arc".  This way you could choose the portion of your playlist that matches your current mood the best, similarly to the My Mixes but with your preselected songs in a continuous manner.

(My Mixes is an excellent feature, by the way!)
![image](https://user-images.githubusercontent.com/910321/122558422-5a82bf80-d03e-11eb-9076-737f981c2cc2.png)

**Describe the solution you'd like**
I would imagine a sorting option in my playlists with the "sort by mood" or "sort by genre" option.

**Describe alternatives you've considered**
It is now possible to do this manually, but it's a lot of time and continuous management. It's not practical.

**Additional context**
I use ytmdesktop v1.13.0 on Windows 10.

**Comment by flleeppyy** [comment]
I've read this multiple times now and I don't think this would be possible with our current team of developers, especially with how slow development has gone (I'm hoping I can pick up this project more since I have more time on my hands).
I feel it would require AI to determine a songs mood then figure out which song should be the first, 2nd, 3rd, etc, based on the energy it has, or whichever style you choose to sort from. 
it's an awesome idea though and I hope I'm not interpreting your idea incorrectly.

**Comment by andrasfuchs** [comment]
Yes, I think it's not trivial, but there are open source genre classification models like this available: 
https://towardsdatascience.com/music-genre-classification-with-tensorflow-3de38f0d4dbb
https://github.com/celestinhermez/music-genre-classification

This one is a huge collection of genres and some sample songs:
https://everynoise.com/

As a first step it would be really fun to show in ytmdesktop what the AI thinks the genre of a song is. The users could agree or disagree and we could use that feedback to train the AI model more and more. 
Is it public information how many of us use ytmdesktop?

I know I simplified the problem and the devil is in the details, but still, it's a exciting path to discover ;)

I ping the author of the above genre classification repo @celestinhermez, hoping that he finds this idea also interesting.

**Comment by adlerluiz** [comment]
I don't think it would be possible to make this classification so easily just having information from the client, but it can be studied for a future implementation.

I've been away for a while for some problems, but I'm back on the project and I'm seeing a lot of people helping and creating new things (unlike the beginning, where there were only requests for improvements and features :P ).

There are currently about 110k users using YTMDesktop

**Comment by andrasfuchs** [comment]
First of all congratulations for all the contributors for the 110k users, that's great! 
With so many active users we could train a decent AI model to detect music styles properly.

My theory is that we could use a dataset like everynoise.com has, and fine tune it later based on user feedback. 
I'll contact the authors of Every Noise hoping that they could provide us an API to access their genre-map.

As a start I'll also look into ytmdesktop's source code. I think it would be great to have genre prediction values in the _Stats for nerds_ box in the beginning, and add the user feedback controls later.
![ytmdesktop_issue745a](https://user-images.githubusercontent.com/910321/124256936-232b0d00-db2c-11eb-97bd-314f10d69dc1.png)

I'd like to contribute my time to this feature somehow, so let me know if you see a way that would make the task easier.

**Comment by adlerluiz** [comment]
I'm totally open for its implementation, as you want to contribute to the project, it would be really good to take a look at the YTMDesktop code and any questions to ask me or any contributor so we can help you, there is also our discord server so you can take it out any doubt

**Comment by andrasfuchs** [comment]
I've just got a reply from Glenn, the author of Every Noise and his website uses Spotify API. 
As he put it, the API's _"terms of use would prevent you from using a Spotify API to power a YouTube feature"_. 
I wonder if it actually prevents YTMDesktop too, since its not affiliated with Google/YouTube. Do you know if this is the case?

Thanks for the Discord invitation, I'll join as soon as I have spent enough time with the source code to have any questions :)

**Comment by andrasfuchs** [comment]
I spent part of the weekend reading and testing different approaches to this challenge. Bear in mind, that I'm not an machine learning expert, I'm just excited about this feature.

Automatic AI-based genre-definition is already exists and there are some good papers and articles on the subject. 
If you are curious these are good, and fairly recent ones:

- [Music Genre Classification With TensorFlow](https://towardsdatascience.com/music-genre-classification-with-tensorflow-3de38f0d4dbb)
- [How Spotify Uses Machine Learning Models?](https://www.analyticssteps.com/blogs/how-spotify-uses-machine-learning-models)
- [Music Genre Classification Using Machine Learning](https://www.analyticssteps.com/blogs/music-genre-classification-using-machine-learning)
- [Meet the man classifying every genre of music on Spotify — all 1,387 of them](https://www.thestar.com/entertainment/2016/01/14/meet-the-man-classifying-every-genre-of-music-on-spotify-all-1387-of-them.html)

Technically speaking we would need to run a machine learning (ML) model with JavaScript on the client side. I've found [this TensorFlow JavaScript library](https://www.tensorflow.org/js), so theoretically it is possible to implement.

I was also looking for good music genre-classification ML models, but as far as I can tell, we would need to develop our own. It is not ideal, but it has a big potential. In practice that would mean that we would need to develop and run a small backend service with API access for the YTMDesktop client. The primary function of this service would be that it would generate questions related to the songs that our users would voluntarily answer every now and then. 

So, for example: the YTMDesktop client would show a question to the listener like "How much do you think this song is 'classical'?". The answers would range from 'Not at all' to 'Totally classical' with a few steps inbetween, numerically from 0 to 1. We would store the answers in our backend database.
The nice potential of this is that we could not just ask about genres (e.g. pop, rock, electronic, etc.), but subgenres (e.g. minimal techno, contemporary jazz, etc.) and moods (melancholic, uplifting, etc.) too.
This would let us develop new features like _generating mood changer radio station_ that starts with one mood and gradually moves to another (e.g. from dark to contempt) or _starting DJ sets_ that would mix similar songs with the same subgenre together without the usual gap between songs.

After we collected enough replies for a particular keyword/tag we could train our own ML model for that keyword. As I understand it now, there is a known way to do it, namely by generating histograms of the song and feed those images paired with their tag values into a ML training software (like TensorFlow or ML.NET).

So this is the bird's eye view that I have now of this task. The main question is for me is: does it worth it? I've been working on backend systems in the past 15 years, so I could gladly help with that, but right now I'm not convinced we would benefit enough to justify the time-investment. If it works we could publish our regularly updated models for the tags (genres and moods) to help the open-source community, or make a separate genre-definition service if we feel like it. @adlerluiz what do you think?

As an alternative I found a wonderful service (of a startup based in Germany) that have API access and does all that we need and more: [Cyanite](https://www.cyanite.ai). I highly recommend you to play around with it. I tested it myself and it works great, I think. I wonder if we could make a cooperation agreement with them and spare the work. I'm not affiliated with them, I found them 2 days ago, I just like what they are doing.

That's all I have now, let me know what you think!

**Comment by adlerluiz** [comment]
I believe people who like to have this control and data available will love it, but that's what you said, we have to measure to see if it's worth all this effort for just a few users to use, we have a good and growing user base that if they always answered the questions, it would help a lot, but I don't know to what extent that will happen.

I agree with the part of doing something to help the open source community and even music lovers, we can also see the possibility of creating as a separate project with API in which YTMDesktop consumes and interacts (something like last.fm in the sense of make a post with current music, questions to help feed the ML model).

If the Cyanite agree to make a deal with us, it would really save a lot of work and we would have instant access to this base, no problem for me, I am in favor of everything that can be created to help, but it will also depend on whether they have this interest.

**Comment by andrasfuchs** [comment]
Alright then, I have just contacted Cyanite, I wonder what they think and how open they are to a potential long-term cooperation.

I'm glad you mentioned [last.fm](https://www.last.fm)! I used it a long time ago, and I haven't checked them since then. They improved a lot, and as I browsed their API documentation, the tagging feature look very promising for song genre definition and categorization. Check [the track.getTopTags API call](https://www.last.fm/api/show/track.getTopTags) for details. This approach would obviously skip the ML part we mentioned earlier and I don't know how much exactly their [API Terms of Service](https://www.last.fm/api/tos) would allow us to do.

Do you see using their API a valid route to implement this feature? If so, would you like me to contact them and ask if (1) we can use their API (2) use the tag data to order our playlists (3) store the tags locally in the client (4) store and process the tags on our backend (if we had one) to train ML models?

**Comment by andrasfuchs** [comment]
I made a quick mock-up of the idea of the user feedback dialog. I would imagine to ask these questions only to a small portion of users, and only for certain tags and certain, popular songs at the beginning. The dialog would disappear automatically after 5-10 seconds, so that it doesn't annoy people.

![YTMDesktop_feedback_mockup](https://user-images.githubusercontent.com/910321/127114948-d63a2cd2-4f05-47bb-a4f5-6187de280e45.png)

The _relaxed_ in the picture could be replaced by any mood or genre tag name.
As we discussed before, this would need a simple backend with API, but I can help with that.
